devito/types/grid.py

```diff
@@ -907,6 +908,180 @@ def bounds(self):
+          [1, 1, 0, 0, 0, 1, 1],
+          [1, 1, 0, 0, 0, 1, 1],
+          [1, 1, 0, 0, 0, 1, 1],
+          [1, 1, 0, 0, 0, 1, 1],
+          [1, 1, 0, 0, 0, 1, 1],
+          [1, 1, 0, 0, 0, 1, 1]], dtype=int32)
+
+    """
+
+    DimSpec = None | dict[Dimension, Dimension | str]
+    ParsedDimSpec = frozendict[Dimension, Dimension | str]
+
+    def __init__(self, grid: Grid, border: int | np.integer,
+                 dims: DimSpec = None, name: str = 'border') -> None:
+
+        self._name = name
+        self._border = border
+        self._border_dims = Border._parse_dims(dims, grid)
+
+        ndomains, bounds = self._build_domains(grid)
+        super().__init__(N=ndomains, bounds=bounds, grid=grid)
+
+    @property
+    def border(self):
+        return self._border
+
+    @property
+    def border_dims(self):
+        return self._border_dims
+
+    @property
+    def name(self):
+        return self._name
+
+    @staticmethod
+    def _parse_dims(dims: DimSpec, grid: Grid) -> ParsedDimSpec:
+        if dims is None:
+            _border_dims = {d: d for d in grid.dimensions}
+        elif isinstance(dims, Dimension):
+            _border_dims = {dims: dims}
+        elif isinstance(dims, dict):
+            _border_dims = dims
+        else:
+            raise ValueError("Dimensions should be supplied as a single dimension, or "
+                             "a dict of the form `{x: x, y: 'left'}`")
+
+        return frozendict(_border_dims)
+
+    def _build_domains(self, grid: Grid) -> tuple[int, tuple[np.ndarray]]:
+        """
+        Constructs the bounds and ndomains kwargs for the SubDomainSet.
+        """
+
+        domain_map = {}  # Stores the side
+        interval_map = {}  # Stores the mapping from the side to bounds
+
+        # Unpack the user-provided specification into a set of sides (on which
+        # a cartesian product is taken) and a mapper from those sides to a set of
+        # bounds for each dimension.
+        for d, s in zip(grid.dimensions, grid.shape):
+            if d in self.border_dims:
+                side = self.border_dims[d]
+
+                if isinstance(side, Dimension):
+                    domain_map[d] = (LEFT, CENTER, RIGHT)
+                    interval_map[d] = {LEFT: (0, s-self.border),
+                                       CENTER: (self.border, self.border),
+                                       RIGHT: (s-self.border, 0)}
+                elif side == 'left':
+                    domain_map[d] = (LEFT, CENTER)
+                    interval_map[d] = {LEFT: (0, s-self.border),
+                                       CENTER: (self.border, 0)}
+                elif side == 'right':
+                    domain_map[d] = (CENTER, RIGHT)
+                    interval_map[d] = {CENTER: (0, self.border),
+                                       RIGHT: (s-self.border, 0)}
+                else:
+                    raise ValueError(f"Unrecognised side value {side}")
+            else:
```

Comment:
how do you end up here?

Reply:
Oh wait, this lower section? You enter that `else` clause if the user omits a dimension from the mapper provided, so something like `{y: y}` will result in no border in the `x` dimension